[Hoang Le Tri Cuong] iP

ErnestCuong opened the pull request:
### DukePro Max Plus Ultra
> "Imagination is more important than knowledge" - Albert Einstein ([source](https://www.inc.com/annabel-acton/10-einstein-quotes-to-fire-up-your-creativity.html#:~:text=1.,Imagination%20encircles%20the%20world.%22))

DukePro Max Plus Ultra allows you to track your tasks with ease. Or is it? ([Vsauce](https://www.youtube.com/watch?v=IfX1hJS0iGM&t=37s))

This chat box is,
* text-based
* FAST FAST FAST
* ~~EASY~~ SUPERCALIFRAGILISTICEXPIALIDOCIOUSLY EASY to use

All you need to do is,
1. Download DukePro Max Plus Ultra [here](https://www.youtube.com/watch?v=dQw4w9WgXcQ) :wink:
2. Double-click it
3. Add your tasks
4. Magic!

Wait, did I forget to mention the best part? It's **FREE**! :partying_face:

Currently we have the following features:
- [x] Task Management
- [ ] Reminders (Coming soon!)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

ErnestCuong committed:
Add Level 1

ErnestCuong committed:
Add Level 2

ErnestCuong committed:
Add Task class

ErnestCuong committed:
Add Level 3

ErnestCuong committed:
Level-3-v2

ErnestCuong committed:
Add Event, ToDo, Deadline Classes and update Duke to Level 4

ErnestCuong committed:
Add DukeException

ErnestCuong committed:
Add Automated Text UI Testing

ErnestCuong committed:
Update DukeException

ErnestCuong committed:
Add Level 5 and Update EXPECTED.TXT

ErnestCuong committed:
Update 'list' command for Duke and EXPECTED.TXT

ErnestCuong committed:
Update Duke to use ArrayList

ErnestCuong committed:
Add 'delete' command for Duke.java

ErnestCuong committed:
Update Test (input.txt and EXPECTED.TXT)

ErnestCuong committed:
Branch Level 7

ErnestCuong committed:
Add Level 8

ErnestCuong committed:
Update input.txt and EXPECTED.TXT for Level 8

ErnestCuong committed:
Merge branch 'branch-Level-7'

ErnestCuong committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java

ErnestCuong committed:
Add Command

ErnestCuong committed:
Add OOP

ErnestCuong committed:
Update Duke.java and EXPECTED.TXT

ErnestCuong committed:
Add Package

ErnestCuong committed:
Fix Path

ErnestCuong committed:
Add JUnit Tests

ErnestCuong committed:
Add Level 9 Find Command

ErnestCuong committed:
Merge branch 'branch-Level-9'

ErnestCuong committed:
Add JavaDoc

ErnestCuong committed:
Add JavaDoc

ErnestCuong committed:
Add JavaDoc from JavaDoc branch

ErnestCuong committed:
Merge branch 'branch-A-JavaDoc'

ErnestCuong committed:
Apply Coding Standard to Ui Class

ErnestCuong committed:
Merge branch 'branch-A-CodingStandard'

ErnestCuong committed:
Merge remote-tracking branch 'origin/add-gradle-support'

ErnestCuong committed:
Change mainClassName in build.gradle

ErnestCuong committed:
Add CheckStyle

ErnestCuong committed:
Add FXML implementation

ErnestCuong committed:
Integrate Duke response into the GUI

ErnestCuong committed:
Merge branch 'branch-Level-10'

rebeccacxy commented on src/main/java/duke/FindCommand.java:
```diff
@@ -0,0 +1,31 @@
+            TaskList found = new TaskList();
+            for (int i = 0; i < tasks.numOfTasks(); i++) {
+                Task curr = tasks.getTask(i);
+                if (curr.checkDescription(userInput.substring(5))) found.addTask(curr);
```
Should conditionals be put on a separate line and wrapped by curly brackets?

ErnestCuong replied:
Yes, you are right. Thank you for pointing this out.

rebeccacxy commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,86 @@
+                                    DateTimeFormatter.ofPattern("MMM dd yyyy HH:mm"))
+                    );
+                    tasks.add(currentTask);
+                } else {
+                    // do nothing
+                }
```
Is this best practice to have an empty else clause? Perhaps an exception could be thrown here

ErnestCuong replied:
You are right, I should throw an exception here.

rebeccacxy commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,27 @@
+package duke;
+
+public class Task {
+    protected String description;
+    protected boolean isDone;
+
+    public Task(String description) {
+        this.description = description;
+        this.isDone = false;
+    }
+
+    public boolean checkDescription(String input) {
+        return description.contains(input);
+    }
+
+    public String getStatusIcon() {
+        return (isDone ? "X" : " "); // mark done task with X
+    }
+
+    public void markAsDone() {
+        isDone = true; // mark this task as done
+    }
```
Missing javadocs for this task class?

ErnestCuong replied:
Yes, thank you for pointing out.

rebeccacxy commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,46 @@
+
+import java.util.ArrayList;
+
+public class TaskList {
+    private ArrayList<Task> tasks;
+
+    public TaskList(ArrayList<Task> storage) {
+        this.tasks = storage;
+    }
+
+    public TaskList() {
+        this.tasks = new ArrayList<>();
+    }
+
+    public void addTask(Task task) {
+        this.tasks.add(task);
+    }
+
+    public Task getTask(int i) {
+        return tasks.get(i);
+    }
+
+    public Task removeTask(int i) {
+        return this.tasks.remove(i);
+    }
+
+    public int numOfTasks() {
+        return tasks.size();
+    }
```
Missing javadocs for this tasklist class?

ErnestCuong replied:
Yes, thank you for pointing this out. I will update them as soon as possible. I was in a rush to submit so I forgot to put javadocs for some classes and methods.

rebeccacxy commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,23 @@
+package duke;
+
+public class Parser {
+    public Parser() {
+
+    }
```
Is this empty constructor needed?

ErnestCuong replied:
I was not aware that there is no need to include an empty constructor. Thank you for pointing this out.

rebeccacxy commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,86 @@
+    Path currentRelativePath = Paths.get("");
+    String s = currentRelativePath.toAbsolutePath().toString();
+
+    public Storage() {
+
+    }
```
Is this empty constructor needed?

rebeccacxy commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,86 @@
+
+    }
+
+    public ArrayList<Task> load() throws DukeException {
```
Since your save method is named `saveTasks`, would it be better to name this `loadTasks`? For standardisation

ErnestCuong replied:
Yes, you are right.

rebeccacxy commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,61 @@
+    public Ui() {
+    }
+
+    public String showExitMessage() {
+        return "Bye. Hope to see you again soon!";
+    }
+
+    public String showLoadingError() {
+        return "There is a problem loading saved data.";
+    }
+    
+    public String showAddTask(Task task, int numOfTasks) {
+        return "Got it. I've added this task:\n  "
+                + task.toString() + "\nNow you have " + numOfTasks
+                + " task(s) in the list.";
+    }
+
+    public String showDeleteTask(Task task, int numOfTasks) {
+        return "Noted. I've removed this task:\n  "
+                + task.toString() + "\nNow you have " + numOfTasks
+                + " task(s) in the list.";
+    }
+
+    public String showTaskList(TaskList tasks) {
+        return "Here are the tasks in your list:" + tasks.toString();
+    }
+
+    public String showFoundTasks(TaskList tasks) {
+        return "Here are the matching tasks in your list:" + tasks.toString();
+    }
+
+    public String showTaskDone(Task task) {
+        return "Nice! I've marked this task as done:\n  "
+                + task.toString();
+    }
+
+    public String showError(String s) {
+        return s;
+    }
```
Would it be better to name these methods `getXXX` instead of `showXXX`?

ErnestCuong replied:
Yes, I initially put the methods as `void` but after changing them to return `String` I didn't change the name. Thank you for the suggestion.

rebeccacxy commented on src/main/java/MainWindow.java:
```diff
@@ -0,0 +1,54 @@
+    public void initialize() {
+        scrollPane.vvalueProperty().bind(dialogContainer.heightProperty());
+
+        String greeting = "Hello! I'm Duke\nWhat can I do for you?";
```
Would it be better for this to be a global variable? And naming in caps, ie `private static final String GREET = ...` And also would it be better for this to be done in the Ui class and set a getter to it?

ErnestCuong replied:
You are right. Your suggestion seems to be better.

rebeccacxy reviewed (commented):
Overall, code quality is pretty good and I didn't find many violations. 
However, you are missing javadocs in many classes (all public methods must have them iirc) and there are some coding standard violations you can take note of. It is also better to put your classes in separate packages, ie. `storage`, `tasks`, `util` instead of all in one `duke` package. 

Good job and all the best for the rest of IP!

jayasting98 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,58 @@
+    }
+
+
+    public void run() {
```
Perhaps you can consider adding a Javadoc header comment for this public method. Otherwise, maybe you can also consider making it private since, if I am not wrong, I do not think this is called anywhere else.

jayasting98 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,58 @@
+import duke.TaskList;
+import duke.Ui;
+
 public class Duke {
```
Should this have a Javadoc header comment as well since it is a public class?

ErnestCuong replied:
Yes, thank you for pointing this out. I will update them as soon as possible. I was in a rush to submit so I forgot to put javadocs for some classes and methods.

jayasting98 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,58 @@
+    private TaskList tasks;
+    private Ui ui;
+
+    public Duke() {
```
Maybe you should consider adding a Javadoc header comment for this public constructor too, or make it private since I think it is only called here if I am not mistaken.

jayasting98 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,58 @@
+        ui.closeUserInput();
+    }
+
     public static void main(String[] args) {
```
Maybe you should add a Javadoc header comment for this public method too, though perhaps it is not that necessary for the main method.

jayasting98 commented on src/main/java/duke/AddCommand.java:
```diff
@@ -0,0 +1,79 @@
+        try {
+            Task curr = new ToDo(s.substring(5));
+            tasks.addTask(curr);
+            return ui.showAddTask(curr, tasks.numOfTasks());
+        } catch (IndexOutOfBoundsException e) {
+            throw new DukeException("The description of a todo cannot be empty.");
+        }
+    }
+
+    public static String addEventResponse(String s, TaskList tasks, Ui ui) throws DukeException {
+        try {
+            int at = s.lastIndexOf(" /at ");
+            Task curr = new Event(
+                    s.substring(6, at),
+                    LocalDateTime.parse(s.substring(at + 5), DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm"))
+            );
+            tasks.addTask(curr);
+            return ui.showAddTask(curr, tasks.numOfTasks());
+        } catch (IndexOutOfBoundsException e) {
+            throw new DukeException("The description and time of an event cannot be empty.");
+        } catch (DateTimeParseException e) {
+            throw new DukeException("The time must be in this format: yyyy-MM-dd HH:mm");
+        }
+    }
+
+    public static String addDeadlineResponse(String s, TaskList tasks, Ui ui) throws DukeException {
+        try {
+            int by = s.lastIndexOf(" /by ");
+            Task curr = new Deadline(
+                    s.substring(9, by),
+                    LocalDateTime.parse(s.substring(by + 5), DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm"))
+            );
+            tasks.addTask(curr);
+            return ui.showAddTask(curr, tasks.numOfTasks());
+        } catch (IndexOutOfBoundsException e) {
+            throw new DukeException("The description and time of a deadline cannot be empty.");
+        } catch (DateTimeParseException e) {
+            throw new DukeException("The time must be in this format: yyyy-MM-dd HH:mm");
+        }
+    }
```
Perhaps you should consider adding Javadoc header comments for this public class and public members as well. I noticed you were missing Javadoc header comments for other public things as well.

jayasting98 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,86 @@
+import java.util.Scanner;
+
+public class Storage {
+    Path currentRelativePath = Paths.get("");
+    String s = currentRelativePath.toAbsolutePath().toString();
```
Is there any reason these are non-private? Does `Storage` have subclasses that require access to these? On the other hand, even if there are, perhaps you could consider using getter methods even for only subclasses to use.

jayasting98 replied:
Since `s` is a variable with a large scope (a class instance variable), maybe consider using a longer, more descriptive name for it, or perhaps consider moving it to a smaller scope.

ErnestCuong replied:
Yes, I forgot to put the access modifier `private`. And you suggestion for the naming is correct. However, these fields are not really used anywhere, so I probably should delete them. Thank you for the suggestions.

jayasting98 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,46 @@
+public class TaskList {
+    private ArrayList<Task> tasks;
+
+    public TaskList(ArrayList<Task> storage) {
+        this.tasks = storage;
+    }
+
+    public TaskList() {
+        this.tasks = new ArrayList<>();
+    }
+
+    public void addTask(Task task) {
+        this.tasks.add(task);
+    }
+
+    public Task getTask(int i) {
+        return tasks.get(i);
+    }
+
+    public Task removeTask(int i) {
+        return this.tasks.remove(i);
+    }
```
Since there are no other similarly-named local variables as `tasks` in some of these, perhaps you could consider removing `this` from `this.tasks`.

ErnestCuong replied:
Noted. Thank you for this suggestion.

jayasting98 reviewed (commented):
Overall, I found your code had minimal coding style violations, except maybe missing Javadoc header comments in some places. Still, it is pretty clean overall. Keep it up!

ErnestCuong committed:
Add Assertions to Storage.java and Ui.java

Use Assertions to keep track of assumptions made in Storage and Ui.

These assumptions are:
* Deadline task stored has "by" and Event task stored has "at"
* Number of tasks is never negative

The purpose of using these Assertions is to alert if future changes invalidate any of these assumptions.

ErnestCuong committed:
Merge pull request #1 from ErnestCuong/branch-A-Assertions

Add Assertions to Storage.java and Ui.java

ErnestCuong committed:
Add dummy commit

ErnestCuong committed:
Add missing JavaDocs

There were missing Javadocs

They are now all added except for JUnit test classes

ErnestCuong committed:
Merge branch 'branch-A-CodeQuality'

ErnestCuong committed:
Add first version of Code Quality Improvement

Storage class method `load` was previously too long

The parsing of stored data is extracted out into a helper function called `parseStoredTask`

ErnestCuong committed:
Merge branch 'branch-A-CodeQuality'

ErnestCuong committed:
Add doAfterTask and doAfterDateTime fields for Task class

It is necessary to track when a task needs to be done after a specific task/time

Therefore these 2 fields are added into the Task class

This would enable B-DoAfterTasks to be implemented

ErnestCuong committed:
Add B-DoAfterTasks first version

This supports setting a task after another task or after a certain date

However this is cannot currently be saved into storage yet.

ErnestCuong committed:
Merge branch 'branch-B-DoAfterTasks'

ErnestCuong committed:
Add Better GUI

ErnestCuong committed:
Update Duke to store DoAfterTasks information

ErnestCuong committed:
Add icon

ErnestCuong committed:
Refine code quality for Storage class

ErnestCuong committed:
Add icon

ErnestCuong committed:
Update docs README.md partial

ErnestCuong committed:
v2

ErnestCuong committed:
v3

ErnestCuong committed:
Update Final

ErnestCuong committed:
Merge pull request #3 from ErnestCuong/docs-readme

ErnestCuong committed:
Update README.md

ErnestCuong committed:
Add typo fix in Ui.java

ErnestCuong committed:
Fix bugs for storing of DoAfterTask

ErnestCuong committed:
Update README.md

ErnestCuong committed:
Merge pull request #4 from ErnestCuong/docs-readme

Update README.md

ErnestCuong committed:
Update README.md

ErnestCuong committed:
Update README.md

ErnestCuong committed:
Fix user input splitting bugs in AddCommand.java

ErnestCuong committed:
Merge branch 'master' of https://github.com/ErnestCuong/ip

ErnestCuong committed:
Fix header comments

ErnestCuong committed:
Fix Storage bug

ErnestCuong committed:
Change after command messages

ErnestCuong committed:
Update README.md

ErnestCuong committed:
Update README.md

ErnestCuong committed:
Update README.md

ErnestCuong committed:
Update README.md

ErnestCuong committed:
Fix for checkstyle

ErnestCuong committed:
Add Duke test method

ErnestCuong committed:
Merge branch 'master' of https://github.com/ErnestCuong/ip

ErnestCuong committed:
Update error message for after and do after commands
